File: daily coding/reach_n_to_1.py
"""
This problem was asked by PagerDuty.

Given a positive integer N, find the smallest number of steps it will take to reach 1.

There are two kinds of permitted steps:

You may decrement N to N - 1.
If a * b = N, you may decrement N to the larger of a and b.
For example, given 100, you can reach 1 in five steps with the following route: 100 -> 10 -> 9 -> 3 -> 2 -> 1.

"""
# Given a positive integer N, find the smallest number of steps it will take to reach 1
# There are two kinds of permitted steps:
# You may decrement N to N - 1.
# If a * b = N, you may decrement N to the larger of a and b.
# For example, given 100, you can reach 1 in five steps with the following route: 100 -> 10 -> 9 -> 3 -> 2 -> 1.

def issquare(n):
    return n**0.5 == int(n**0.5)
N = 10
steps = 0
print(f'{N} >> ', end='')
while N > 1:
    if N % 2 == 0:
        N = N // 2
    else:
        N = N - 1
    steps += 1
    print(f'{N} >> ', end='')
print(f'\n{steps=}')

File: daily coding/reach_n_to_1.c

// Given a positive integer N, find the smallest number of steps it will take to reach 1

// You can perform any one of the following 3 steps

// 1.) Subtract 1 from it. ( n = n - 1 )

// 2.) If its divisible by 2, divide by 2. ( if n % 2 == 0 , then n = n / 2 )

// 3.) If its divisible by 3, divide by 3. ( if n % 3 == 0 , then n = n / 3 )

#include <stdio.h>

int min(int a, int b)
{
    return a < b ? a : b;
}

int reach_n_to_1(int n)

{   
    if (n == 1)
        return 0;
    int res = reach_n_to_1(n - 1);
    if (n % 2 == 0)
        res = min(res, reach_n_to_1(n / 2));
    if (n % 3 == 0)
        res = min(res, reach_n_to_1(n / 3));
    return res + 1;
}


int main()
{
    int n = 100;
    printf("Minimum steps required to reach 1 from %d is %d", n, reach_n_to_1(n));
    return 0;
}